### ISSUE TITLE
How to get socket.io variable from server-side page?

### ISSUE THREAD

**Issue by jolacaleb** [meanjs/mean#1114]
Hello.

I'm trying to use socket.io to send a message to specific client when there's update to database.

so I tried to write code after user.save, like below..

```javascript
exports.updateUser = function(req, res) {
// .... Find user by findById, and do something... and, 
user.save(function(err) {
        if (err) {
          res.status(400).send({
            message: err
          });
        }
        else {
          // i need to send a message to a specific client here...
          res.json(200);
        }
      });
}
```
But I can't get socket.io object to emit.
How can I get socket.io object to emit or broadcast message there?
Any help would be appreciated, thank you!

**Comment by lirantal** [comment]
@jolacaleb you can take a look at the chat example which is implemented in the project.

**Comment by mleanos** [comment]
@jolacaleb The current implementation of the Chat feature in this project, provides a great example of using socket.io to communicate between the server & client; just as Liran suggested. However, it is a little limiting for certain use cases (or project requirements). The server socket configuration in this project assumes that the *emit* is being initiated from the client-side. It sounds like your need is to have the *emit* coming from a server-side controller. If that's the case then you have a few options; although, I would recommend weighing your options out before choosing which route to take. It's been quite an endeavor for me to figure out the best way to utilize Socket.IO to satisfy my application's requirements. I still don't know which is best. I think you just have to figure out what makes sense for you.

tl; dr;
Options:
1) Stick with the client/server implementation currently used in this project. 
2) Create a global variable in express to hold your Socket.IO object, that you can reference in any part of your server-side. 
  This article explains this concept: https://vexxhost.com/resources/tutorials/mean-socket-io-integration-tutorial/
3) Have a look at this POC I was playing around with. This is a bit different from what you probably need. However, you may find it useful. https://github.com/mleanos/mean/commit/a1a869dafda3d24b641209fae4c207f6485aee96

You have to find what works for you. You may be able to find an interesting way to combine the implementations of (2) & (3). Report back with what your end up going with. I'm always curious to know how others are using Socket.IO.

**Comment by mleanos** [comment]
Also, https://github.com/meanjs/mean/issues/792

**Comment by jolacaleb** [comment]
Thanks for the comments!

I solved this problem with @mleanos number 2 option,

First, I set socket.io as global app in express,
```javascript
// config/lib/socket.io.js
{
...
  // Add an event listener to the 'connection' event
  io.on('connection', function (socket) {
    config.files.server.sockets.forEach(function (socketConfiguration) {
      require(path.resolve(socketConfiguration))(io, socket);
    });
  });

// added this line below
app.set('socketio', io);

return server;
}
```
Second, I load this variable as 'app.get' which is like snipplet below:
```javascript
// any server controller file, I want to emit, 
// in my case, user.profile.server.controller.js
{
var io = req.app.get('socketio');

...
io.to(user._id).emit('message', msg);
}
```
Works as I expected! Thank you for all comments!

**Comment by mleanos** [comment]
@jolacaleb I'm glad that worked out for you. Thanks for reporting back!

**Comment by mleanos** [comment]
Closing, since @jolacaleb reported back that the problem was solved.

**Comment by saintwilbur** [comment]
In the same vein, what is the best practice to call functions belonging to the *.server.controllers on socket events?
The file module/*/server/sockets/*.server.sockets.js is a good place to define what to do on the various emit messages. Should I be referencing the controller here, to call it's functions?
This does not seem like the right place to define functions that should belong to the controller.

**Comment by nidhalmesselmani** [comment]
thank you so much

**Comment by frkwhiteangel** [comment]
```
   ReferenceError: app is not defined

        var io = req.app.get('socketio');
```

any idea ?

**Comment by nehemie054** [comment]
frkwhiteangel, I think that might be due to that you attached the app after importing the controllers or other files accessing the global app.set instance.
First set the **app.set('socketio', io)**
and then import other stuff later.
